spec/unit/agecounters.spec.js

```diff
@@ -0,0 +1,133 @@
+const AgeCounters = require("../../lib/components/agecounters");
+
+describe("AgeCounters", function() {
+    describe("constructor", function() {
+        it("Can construct some counter periods by default", function() {
+            const ageCounter = new AgeCounters();
+            expect(ageCounter.counterPeriods).toEqual(["1h", "1d", "7d", "all"]);
+            expect(ageCounter.counters.size).toEqual(4);
+            const mapIter = ageCounter.counters.keys();
+            expect(mapIter.next().value).toEqual(3600);
+            expect(mapIter.next().value).toEqual(3600 * 24);
+            expect(mapIter.next().value).toEqual(3600 * 24 * 7);
+            expect(mapIter.next().value).toEqual("all");
+        });
+
+        it("Can construct given counter periods", function() {
+            const ageCounter = new AgeCounters(["1h", "2d", "5d", "3w"]);
+            expect(ageCounter.counterPeriods).toEqual(["1h", "2d", "5d", "3w", "all"]);
+            expect(ageCounter.counters.size).toEqual(5);
+            const mapIter = ageCounter.counters.keys();
+            expect(mapIter.next().value).toEqual(3600);
+            expect(mapIter.next().value).toEqual(3600 * 24 * 2);
+            expect(mapIter.next().value).toEqual(3600 * 24 * 5);
+            expect(mapIter.next().value).toEqual(3600 * 24 * 7 * 3);
+            expect(mapIter.next().value).toEqual("all");
+        });
+
+        it("Can construct with an empty array", function() {
+            const ageCounter = new AgeCounters([]);
+            expect(ageCounter.counterPeriods).toEqual(["all"]);
+            expect(ageCounter.counters.size).toEqual(1);
+        });
+        /* eslint-disable no-new */
+        it("Cannot construct with invalid period strings", function() {
+            expect(() => {new AgeCounters(["cats", "dogs"]);}).toThrow();
+            expect(() => {new AgeCounters(["5"]);}).toThrow();
+            expect(() => {new AgeCounters(["h"]);}).toThrow();
+            expect(() => {new AgeCounters(["1x"]);}).toThrow();
+        });
+        it("Can only construct with positive integers", function() {
+            expect(() => {new AgeCounters(["-1h"]);}).toThrow();
+            expect(() => {new AgeCounters(["0h"]);}).toThrow();
+        });
+
+        it("Cannot construct counter with null", function() {
+            expect(() => {new AgeCounters([null]);}).toThrow();
+            expect(() => {new AgeCounters([undefined]);}).toThrow();
```

Comment:
also explicitly passing `null` and `undefined` in place of the array (literally `new AgeCounters(null)`) - should fall to the defaults.

Reply:
*Surely* I don't need to test undefined since that's the same as passing nothing. I'll test null though :)